File: cs50/week1/cash/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    // User inputs CC number
    long n = get_long("Credit Card Number\n");

    // Initialise power counter
    int i = 0;
    int sum = 0;
    long x = 0;
    int length[16];

    do
    {
        // finding individual digit
        x = ((n / pow(10, i)));
        int y = x % 10;
        length[i] = y;

        if (i % 2 == 0)
        {
            //adding normal digits
            sum = sum + y;
        }
        // adding the sum of the remaining digits
        else
        {
            int z = y * 2;
            int a = ((z / pow(10, 0)));
            int b = ((z / pow(10, 1)));
            sum = sum + a % 10 + b % 10;
        }

        i ++;
    }
    while (x != 0);

    // Checksum mod 10 value
    if ((sum % 10 == 0) && (i == 14 || i == 16 || i == 17))
    {
        if (length[i - 2] == 4)
        {
            printf("VISA\n");
        }
        else if (length[i - 2] == 3 && (length[i - 3] == 4 || length[i - 3] == 7))
        {
            printf("AMEX\n");
        }
        else if (length[i - 2] == 5 && (length[i - 3] == 1 || length[i - 3] == 2 || length[i - 3] == 3 || length[i - 3] == 4
        || length[i - 3] == 5))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: cs50/week6/sentimental-credit/credit.py
#!/usr/local/bin/python
# called a shebang, has to be first line. Directs OS to interpreter
# test cases, optimal way of inputting test cases?

import re

# Protects script if imported from being run
if __name__ == "__main__":
    # get user input
    cc_num = input("Number: ")
    regex = r"^(\d{13}|\d{15,16})$"

    # Validate User input
    if re.match(regex, cc_num) is None:
        print("INVALID")
        exit()

    # find every second number * 2
    # List comphrension
    #digits = [int(x) for x in cc_num[::-1]]
    #y = []
    #for x in range(1, len(cc_num), 2):
        #y = y + [digits[x] * 2]

    # Pythonic way of writing loop, no indexes needed
    for x in digits[1: : 2]:
        y = y + [x * 2]

    # find sum of digits from above
    result = 0
    for elem in y:
        sum_val = 0
        for digit in str(elem):
            sum_val += int(digit)
        result += sum_val

    # add above to sum of remaining digits in cc_num
    #for x in range(0, len(cc_num), 2):
        #result += sum([digits[x]])

    # Pythonic way of writing loop, no indexes needed
    for x in digits[:: 2]:
        result += sum(x)

    # checksum
    if result % 10 != 0:
        print("INVALID")
        exit()

    # output whether valid or not
    if int(cc_num[:2]) == 34 or int(cc_num[:2]) == 37:
        print("AMEX")
    elif int(cc_num[:2]) == 51 or int(cc_num[:2]) == 52 or int(cc_num[:2]) == 53 or int(cc_num[:2]) == 54 or int(cc_num[:2]) == 55:
        print("MASTERCARD")
    elif int(cc_num[:1]) == 4:
        print("VISA")
    else:
        print("INVALID")